packages/core/src/utilities/ProgressiveIterator.ts

```diff
@@ -0,0 +1,169 @@
+export class PromiseIterator<T> extends Promise<T> {
+  iterator?: ProgressiveIterator<T>;
+}
+
+export type ErrorCallback = (message: string | Error) => void;
+
+/**
+ * A progressive iterator is an async iterator that can have data delivered
+ * to it, with newer ones replacing older iterations which have not yet been
+ * consume.  That allows iterating over sets of values and delivering updates,
+ * but always getting the most recent instance.
+ */
+export default class ProgressiveIterator<T> {
+  public done;
+  public name?: string;
+
+  private nextValue;
+  private waiting;
+  private rejectReason;
+
+  constructor(name?) {
+    this.name = name || 'unknown';
+  }
+
+  /** Casts a promise, progressive iterator or promise iterator to a
+   * progressive iterator, creating one if needed to resolve it.
+   */
+  public static as(promise) {
+    if (promise.iterator) {
+      return promise.iterator;
+    }
+    const iterator = new ProgressiveIterator('as iterator');
+    iterator.generate((iterator) => {
+      return promise.then(
+        (v) => iterator.add(v, true),
+        (reason) => iterator.reject(reason)
+      );
+    });
+    return iterator;
+  }
+
+  /** Add a most recent result, indicating if the result is the final one */
+  public add(x: T, done = false) {
```

Comment:
since we are assigning to the `this.nextValue` isn't it like `update` instead of add?

Reply:
maybe not, ignore my comment